optimizer/plan/newplanbuilder.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"fmt"
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+)
+
+// UseNewPlanner means if use the new planner.
+var UseNewPlanner = false
+
+func (b *planBuilder) buildNewSinglePathPlan(node ast.ResultSetNode) Plan {
+	switch x := node.(type) {
+	case *ast.Join:
+		return b.buildNewJoin(x)
+	case *ast.TableSource:
+		switch v := x.Source.(type) {
+		case *ast.SelectStmt:
+			return b.buildSelect(v)
+		case *ast.UnionStmt:
+			return b.buildUnion(v)
+		case *ast.TableName:
+			//todo: select physical algorithm during cbo phase.
+			return b.buildNewTableScanPlan(v)
+		default:
+			b.err = ErrUnsupportedType.Gen("unsupported table source type %T", v)
+			return nil
+		}
+	default:
+		b.err = ErrUnsupportedType.Gen("unsupported table source type %T", x)
+		return nil
+	}
+}
+
+func fromFields(col *ast.ColumnNameExpr, fields []*ast.ResultField) bool {
+	for _, field := range fields {
+		if field == col.Refer {
+			return true
+		}
+	}
+	return false
+}
+
+func extractColumns(expr ast.ExprNode) (cols []*ast.ColumnNameExpr, ok bool) {
```

Comment:
This function only handles a few cases, it doesn't extract all the columns in the expression.